### PR TITLE
Make invalidation be recursive

### LINKED ISSUE #2: Invalidation only does direct dependencies
For our current system it should actually do recursive dependency invalidation. 

Current invalidation matching has run for gazebo-plugins
```
rosbuild@p:~/reprepro_updater$ reprepro -v -b /var/www/repos/building/ listfilter quantal "Package (% ros-hydro-*), ( Depends (% ros-hydro-gazebo-plugins[, ]* ) | Depends (% *ros-hydro-gazebo-plugins) )"
```
No match

A known downstream package still exists.
```
rosbuild@p:~/reprepro_updater$ reprepro -v -b /var/www/repos/building/ listfilter quantal "Package (% ros-hydro-*), ( Depends (% ros-hydro-gazebo-ros[, ]* ) | Depends (% *ros-hydro-gazebo-ros) )"
quantal|main|amd64: ros-hydro-gazebo-ros-pkgs 2.1.4-0quantal-20130715-1116-+0000
quantal|main|amd64: ros-hydro-turtlebot-gazebo 2.0.0-0quantal-20130717-2225-+0000
quantal|main|i386: ros-hydro-gazebo-ros-pkgs 2.1.4-0quantal-20130715-1123-+0000
quantal|main|i386: ros-hydro-turtlebot-gazebo 2.0.0-0quantal-20130717-2238-+0000
```

https://github.com/ros-infrastructure/buildfarm/issues/135

### PR DESCRIPTION
**Pull request by tfoote**
FIxes #2 

New version: http://build.ros.org:8080/job/Irel_import-package/15575/console recursively removing all downstream

Old version: http://build.ros.org:8080/job/Irel_import-package/10771/console removing declared dependencies only

### CODE REVIEW

File: src/reprepro_updater/helpers.py [hunk 1]
```diff
@@ -77,18 +77,59 @@ def _run_update_command(repo_dir, distro, commit):
...
+        return []
+    packages = []
+    for l in output.splitlines():
+        elements = l.split()
+        if len(elements) != 3:
+            print "Expected 3 elements got %s, [%s]" % len(elements), elements
+        packages.append(elements[1])
+    return packages
+
+
```
Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: Can you please move this function to the end of the patch? Currently the order of functions is 3, 1, 2 - 1, 2, 3 would be easier to read.

File: src/reprepro_updater/helpers.py [hunk 2]
```diff
@@ -77,18 +77,59 @@ def _run_update_command(repo_dir, distro, commit):
...
+        'listfilter', distro,
+        "Package (% ros-* ), " +
+        "Architecture (== " + arch + " ), " +
+        "( Depends (% *" + package + "[, ]* ) " +
+        "| Depends (% *" + package + " ) )"]
+
+    try:
+        output = subprocess.check_output(reprepro_command)
+    except subprocess.CalledProcessError as ex:
+        print "Execution of [%s] Failed:" % reprepro_command, ex
```
Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: Please write valid Python 3 code. Same below.

File: src/reprepro_updater/helpers.py [hunk 3]
```diff
@@ -77,18 +77,59 @@ def _run_update_command(repo_dir, distro, commit):
...
+        return []
+    packages = []
+    for l in output.splitlines():
+        elements = l.split()
+        if len(elements) != 3:
+            print "Expected 3 elements got %s, [%s]" % len(elements), elements
```
Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: And this is not a problem but can continue without handling it differently?

Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: Also not enough arguments for format string.

File: src/reprepro_updater/helpers.py (hunk 2 above)
Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: Also not enough arguments for format string.

File: src/reprepro_updater/helpers.py [hunk 4]
```diff
@@ -77,18 +77,59 @@ def _run_update_command(repo_dir, distro, commit):
...
-                                    "( Depends (% *" + package + "[, ]* ) " +
-                                    "| Depends (% *"+package+" ) )"]
-    return try_run_command(invalidate_dependent_command)
+    # storage for recursion
+    processed_packages = []
+    return _invalidate_dependent(repo_dir, distro, arch, package, processed_packages)
+
+
+def _invalidate_dependent(repo_dir, distro, arch, package, processed_packages):
+    """ Implementation of invalidation for recursion"""
```
Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: PEP257.

**Comment by dirk-thomas** [comment]
Beside the comments lgtm.

File: src/reprepro_updater/helpers.py (hunk 2 above)
Comment (tfoote) [inline src/reprepro_updater/helpers.py]: The whole file is using python2x print functions. I switched the exception to be python3 compatible. But it wouldn't be clean to switch the whole file to print_function in this patch.

Comment (dirk-thomas) [inline src/reprepro_updater/helpers.py]: I didn't mean that. I think new code should be written in a Python 3 compatible way. There is simply no need to write code which is Python 2 only.

**Comment by tfoote** [comment]
Successfully ran after fixups from comments: http://build.ros.org:8080/job/Irel_import-package/15581/console

**Pull request merged**